### PR TITLE
Create shodan_host.rb

### PR DESCRIPTION
**Pull request by Natto97**
```
Module options (auxiliary/scanner/http/shodan_host):

   Name           Current Setting  Required  Description
   ----           ---------------  --------  -----------
   Proxies                         no        A proxy chain of format type:host:port[,type:host:port][...]
   RHOSTS                          yes       The target host(s), range CIDR identifier, or hosts file with syntax 'file:<path>'
   RPORT          80               yes       The target port (TCP)
   SHODAN_APIKEY                   yes       The SHODAN API key
   SSL            false            no        Negotiate SSL/TLS for outgoing connections
   VHOST                           no        HTTP server virtual host

msf6 auxiliary(scanner/http/shodan_host) > reload
[*] Reloading module...
msf6 auxiliary(scanner/http/shodan_host) > run
[-] Auxiliary failed: Msf::OptionValidateError One or more options failed to validate: RHOSTS, SHODAN_APIKEY.
msf6 auxiliary(scanner/http/shodan_host) > set RHOSTS 8.8.8.8
RHOSTS => 8.8.8.8
msf6 auxiliary(scanner/http/shodan_host) > set SHODAN_APIKEY xxxxxx
SHODAN_APIKEY => xxxxxx
msf6 auxiliary(scanner/http/shodan_host) > run
[*] Running module against 8.8.8.8

[+] 8.8.8.8:53
[*] Auxiliary module execution completed
```

### CODE REVIEW

**Comment by gwillcox-r7** [comment]
@Natto97 Any update on creating documentation for this PR and updating the module to fix the issues reported by Travis?

**Comment by Natto97** [comment]
@gwillcox-r7   Update file completed https://github.com/rapid7/metasploit-framework/pull/14511

**Comment by gwillcox-r7** [comment]
@Natto97 You do not need to create a separate PR for the documentation file. I have closed the other PR. Please apply your documentation updates from #14511 to your branch `patch-2` instead, and those changes will populate into this PR.

**Comment by Natto97** [comment]
> @Natto97 Any update on creating documentation for this PR and updating the module to fix the issues reported by Travis?

Has been updated, is it like this @gwillcox-r7

**Comment by gwillcox-r7** [comment]
Hello @Natto97. So I was just reviewing this module and I noticed that we already have a module in the framework right now that performs very similar functionality to your module, specifically https://github.com/rapid7/metasploit-framework/blob/c0b42ff7a2a1b677aaba6acd7ebeefddcffdf38e/modules/auxiliary/gather/shodan_search.rb.

That being said during my tests I did notice that the module I linked above will not work on free accounts as the `ip:` filter is a paid filter within shodan.io. Ironically though, the `host` filter is not, for reasons I don't fully understand. 

So with all this being said it seems the best option would be to update the module at https://github.com/rapid7/metasploit-framework/blob/c0b42ff7a2a1b677aaba6acd7ebeefddcffdf38e/modules/auxiliary/gather/shodan_search.rb so that it adds a new check to see if the user specified an IP as the search filter and if they did, then adjust the query format to specifically use the new API format to prevent users having to use the paid `ip:` filter. 

This will allow us to still gather the same information without having similar code spread across multiple modules, thereby reducing code duplication.

In essence we will just be expanding upon the existing module to add in some newer features.

**Comment by bcoles** [comment]
> so that it adds a new check to see if the user specified an IP as the search filter and if they did, then adjust the query format to specifically use the new API format to prevent users having to use the paid ip: filter.

Is there a difference between the `host` and `ip` filters? If so, this check sounds like it will introduce unexpected behavior.

**Comment by gwillcox-r7** [comment]
> > so that it adds a new check to see if the user specified an IP as the search filter and if they did, then adjust the query format to specifically use the new API format to prevent users having to use the paid ip: filter.
> 
> Is there a difference between the `host` and `ip` filters? If so, this check sounds like it will introduce unexpected behavior.

I don't know

> > so that it adds a new check to see if the user specified an IP as the search filter and if they did, then adjust the query format to specifically use the new API format to prevent users having to use the paid ip: filter.
> 
> Is there a difference between the `host` and `ip` filters? If so, this check sounds like it will introduce unexpected behavior.

I'm actually not sure if the `host` filter is a valid filter upon further review. As for the `ip:` filter allowing hostnames, you are correct in that it does not allow hostnames.

**Comment by gwillcox-r7** [comment]
Alright sorry for the confusion. My original idea was to include this into https://github.com/rapid7/metasploit-framework/blob/c0b42ff7a2a1b677aaba6acd7ebeefddcffdf38e/modules/auxiliary/gather/shodan_search.rb. Here is the output from one such run of this module as it stands at the moment:

```
msf6 auxiliary(gather/shodan_search) > run

[*] Total: 2 on 1 pages. Showing: 1 page(s)
[*] Collecting data, please wait...

Search Results
==============

 IP:Port      City  Country        Hostname
 -------      ----  -------        --------
 8.8.8.8:443  N/A   United States  dns.google
 8.8.8.8:53   N/A   United States  dns.google

[*] Auxiliary module execution completed
msf6 auxiliary(gather/shodan_search) > show options

Module options (auxiliary/gather/shodan_search):

   Name           Current Setting                   Required  Description
   ----           ---------------                   --------  -----------
   DATABASE       false                             no        Add search results to the database
   MAXPAGE        1                                 yes       Max amount of pages to collect
   OUTFILE                                          no        A filename to store the list of IPs
   QUERY          ip:8.8.8.8                        yes       Keywords you want to search for
   REGEX          .*                                yes       Regex search for a specific IP/City/Country/Hostname
   SHODAN_APIKEY  *censored* yes       The SHODAN API key

msf6 auxiliary(gather/shodan_search) > 
```
Note that the output is nearly identical to this module, so us merging this module in would not add much net new value minus allowing people to query for the same information using a free Shodan account. 

My thought was to incorporate these changes into https://github.com/rapid7/metasploit-framework/blob/c0b42ff7a2a1b677aaba6acd7ebeefddcffdf38e/modules/auxiliary/gather/shodan_search.rb however it seems it that module was designed from the ground up to incorporate page result support for querying multiple hosts or large data sets at once. Therefore given that the API used within this module does not return any `page` number in its response it would be hard to properly incorporate it into the module.

In conclusion I am in favor of closing this PR out given that we already have similar functionality in the framework and the only thing this really adds in is the ability for free users to obtain the same data automatically.

**Review by gwillcox-r7 — CHANGES_REQUESTED** [review]
Okay very sorry for the confusion here. So what happened was that we were going to try and merge the changes into https://github.com/rapid7/metasploit-framework/blob/c0b42ff7a2a1b677aaba6acd7ebeefddcffdf38e/modules/auxiliary/gather/shodan_search.rb however after a lot of deliberation and analysis we discovered that module heavily relies on the concept of splitting results by pages, which is a pro feature, and the API this module uses doesn't support that feature. Additionally since that module's code is heavily based around the pages feature, its hard to refactor the code to support this new API.

Therefore I think it would be better to leave this module as is and if people have questions as to why some factors might be duplicated we can point them to this discussion. Again sorry for this confusion @Natto97.

Below are a few minor things that I found whilst reviewing the module. I'm going to go ahead and fix these up myself. I'm just leaving them here for your records, however don't worry I'll make sure these are fixed and the module will be landed today once all the changes have been applied and retested to make sure everything still works appropriately.

File: modules/auxiliary/gather/shodan_host.rb [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    if res.code != 200
+      print_error('Shodan did not respond in an expected way. Check your api key')
+    end
+    json = res.get_json_document
+    if json["ports"] != nil
+      json["ports"].each do |post|
+        print_good("#{rhost}:#{post}")
+        report_service(:host => rhost, :port => post, :name => "shodan")
+      end
+    end
```
[the remark this answers is not in the document]
Comment (gwillcox-r7) [reply modules/auxiliary/gather/shodan_host.rb]: Fixed in https://github.com/rapid7/metasploit-framework/pull/14429/commits/12277d3020f8e413b4396f140fe18435e1d6fbf0

File: modules/auxiliary/gather/shodan_host.rb [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+      'uri'    => "/shodan/host/#{rhost}?key=#{key}&minify=true",
+      'method' => 'GET',
+      })
+    res = cli.send_recv(req)
+    cli.close
+    if res.nil?
+      fail_with(Failure::Unknown, 'Unable to connect to shodan')
+    end
+    if res.code != 200
+      print_error('Shodan did not respond in an expected way. Check your api key')
```
[the remark this answers is not in the document]
Comment (gwillcox-r7) [reply modules/auxiliary/gather/shodan_host.rb]: Fixed in https://github.com/rapid7/metasploit-framework/pull/14429/commits/12277d3020f8e413b4396f140fe18435e1d6fbf0

File: modules/auxiliary/gather/shodan_host.rb [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    cli = Rex::Proto::Http::Client.new('api.shodan.io', 443, {}, true)
+    cli.connect
+    req = cli.request_cgi({
+      'uri'    => "/shodan/host/#{rhost}?key=#{key}&minify=true",
+      'method' => 'GET',
+      })
+    res = cli.send_recv(req)
+    cli.close
+    if res.nil?
+      fail_with(Failure::Unknown, 'Unable to connect to shodan')
```
[the remark this answers is not in the document]
Comment (gwillcox-r7) [reply modules/auxiliary/gather/shodan_host.rb]: Fixed in https://github.com/rapid7/metasploit-framework/pull/14429/commits/12277d3020f8e413b4396f140fe18435e1d6fbf0

File: modules/auxiliary/gather/shodan_host.rb [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+  def run
+    # check our API key is somewhat sane
+    unless /^[a-z\d]{32}$/i.match?(datastore['SHODAN_APIKEY'])
+      fail_with(Failure::BadConfig, 'Shodan API key should be 32 characters a-z,A-Z,0-9.')
+    end
+    key = datastore['SHODAN_APIKEY']
+    # Check the length of the key (should be 32 chars)
+    if key.length != 32
+      print_error('Invalid API key (Not long enough)')
+      return
```
[the remark this answers is not in the document]
Comment (gwillcox-r7) [reply modules/auxiliary/gather/shodan_host.rb]: Fixed in https://github.com/rapid7/metasploit-framework/pull/14429/commits/12277d3020f8e413b4396f140fe18435e1d6fbf0

File: modules/auxiliary/gather/shodan_host.rb [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+      )
+    )
+    register_options(
+      [
+        OptString.new('SHODAN_APIKEY', [true, 'The SHODAN API key'])
+      ])
+    deregister_options("SSL")
+    deregister_options("RPORT")
+    deregister_options("VHOST")
+    deregister_options("Proxies")
```
[the remark this answers is not in the document]
Comment (gwillcox-r7) [reply modules/auxiliary/gather/shodan_host.rb]: Fixed in https://github.com/rapid7/metasploit-framework/pull/14429/commits/12277d3020f8e413b4396f140fe18435e1d6fbf0

File: documentation/modules/auxiliary/gather/shodan_host.md [hunk 6]
```diff
@@ -0,0 +1,55 @@
+## Introduction
+This module uses the shodan API to return all port information found on a given host IP.
```
[the remark this answers is not in the document]
Comment (gwillcox-r7) [reply documentation/modules/auxiliary/gather/shodan_host.md]: Fixed in https://github.com/rapid7/metasploit-framework/pull/14429/commits/12277d3020f8e413b4396f140fe18435e1d6fbf0

File: documentation/modules/auxiliary/gather/shodan_host.md [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+  3. Do: `set TARGET <targetip>`
+  4. Do: `set SHODAN_APIKEY <your apikey>`
+  5. Do: `run`
+  6. If the execution is successful, the port opening status of the target server will be obtained
+
+## Options
+
+  **TARGET**
+
+  The target host whose port information needs to be obtained
```
[the remark this answers is not in the document]
Comment (gwillcox-r7) [reply documentation/modules/auxiliary/gather/shodan_host.md]: Fixed in https://github.com/rapid7/metasploit-framework/pull/14429/commits/12277d3020f8e413b4396f140fe18435e1d6fbf0

**Review by gwillcox-r7 — APPROVED** [review]

**Comment by gwillcox-r7** [comment]
# Release Notes
A new module, `auxiliary/gather/shodan_host.rb` has been added which allows users with a Shodan account to enumerate which ports are publicly accessible on a host or set of hosts using Shodan's API.
